### ISSUE TITLE
Framework / support structure for online -> offline routing fallback

### ISSUE THREAD

**Issue by ianthetechie** [stadiamaps/ferrostar#18]
In some scenarios, it would be nice to have a fallback to (or maybe even selectively prefer?) offline routing, such as when the user is in a poor network zone.

Off the top of my head, I'd say maybe we can have a fallback order or something and allow multiple route adapters in sequence. The offline routing would of course be the responsibility of the library end user, as routing itself is out of scope in Ferrostar, but it should support this use case.

**Comment by louwers** [comment]
Why not intergrate OSMR like Organic Maps does? ;-)

**Comment by ianthetechie** [comment]
@louwers is OSMR a project I missed? Or did you mean OSRM?

We already have support for the OSRM data model (it's what both Mapbox and Valhalla [in certain modes] use), and just need to figure out the cleanest+easiest way to integrate offline routing from libosrm or Valhalla without bloating the base library. The data models to support interop are already there :)

**Comment by louwers** [comment]
Typo... 🙈 

> without bloating the base library

Understood.

It would be interesting to have support for completely offline routing. Organic Maps does this extremely well.

**Comment by ianthetechie** [comment]
Yeah, agreed. I really want to build a solid base layer that others (maybe even Organic Maps :D) can hook into. Still a few open questions there, but I'm optimistic :)

btw feel free to open / comment on any other issues as you get ideas! We've got a bit of a priority list/roadmap obviously, but now is totally the time to dream big and collect the ambitious ideas while the codebase is small and flexible.
